Conditional Probabilities for MDPs

tquatmann opened the pull request:
Contains a (re-)implementation of the restart-method for computing conditional probabilities for MDPs.

Also introduces algorithms `bisection`, `bisection-advanced` and `pi` as alternatives for the default restart algorithms. These alternatives work well when the model is acyclic.

Fixes #509 and #734 .

tquatmann committed:
Re-implementation of restart method for conditional probabilities for MDPs

tquatmann committed:
Re-implementation of restart method for conditional probabilities for MDPs

tquatmann committed:
Time measurements for conditional probability computation

tquatmann committed:
Time measurements for conditional probability computation

tquatmann committed:
First version of binary approach

tquatmann committed:
First version of binary approach

tquatmann committed:
Merge remote-tracking branch 'tq-private/conditional' into conditional

tquatmann committed:
bisection approach (simple and advanced version)

tquatmann committed:
policy iteration

tquatmann committed:
Merge remote-tracking branch 'origin/master' into conditional

tquatmann committed:
Merge remote-tracking branch 'origin/master' into conditional

tquatmann committed:
polish conditional implementation

tquatmann committed:
Further polishing conditional probabilities

tquatmann committed:
added a test case for conditional probabilities in mdps

tquatmann committed:
fix formatting

tquatmann commented:
I've added some test cases, so this is now ready to be reviewed :)

tquatmann committed:
Fix issue in restart method

sjunges commented on src/storm/environment/modelchecker/ModelCheckerEnvironment.h:
```diff
@@ -24,6 +25,9 @@ class ModelCheckerEnvironment {
     void setSteadyStateDistributionAlgorithm(SteadyStateDistributionAlgorithm value);
 
+    ConditionalAlgorithm getConditionalAlgorithm() const;
+    void setConditionalAlgorithm(ConditionalAlgorithm value);
```
It doesnt matter performance-wise, but do we pass enums by value or by ref?

tquatmann replied:
We've been doing this mostly by ref in the past, but imo it is not reasonable to do or even enforce doing that:

- Performance-wise, there is no practical difference. Arguably, by-ref is slightly worse as it requires 8 bytes and an additional indirection whereas `enum class`es are essentially `int`s and require 4 bytes (on standard machines).
- Semantics-wise, there is always the chance that unexpected side effects might happen. For example, the reference might get invalidated in the function body because of a vector re-allocation or whatever. Of course, not the case here but generally avoiding references (whenever they are not the intended semantics) means safer code.

tquatmann replied:
Also a good point. There's no silver bullet and I don't think we have been consistent so far.

sjunges commented on src/storm/modelchecker/prctl/SparseMdpPrctlModelChecker.cpp:
```diff
@@ -264,6 +265,8 @@ std::unique_ptr<CheckResult> SparseMdpPrctlModelChecker<SparseMdpModelType>::com
     STORM_LOG_THROW(this->getModel().getInitialStates().getNumberOfSetBits() == 1, storm::exceptions::InvalidPropertyException,
                     "Cannot compute conditional probabilities on MDPs with more than one initial state.");
+    STORM_LOG_THROW(checkTask.isOnlyInitialStatesRelevantSet(), storm::exceptions::InvalidPropertyException,
+                    "Conditional probabilities can only be computed for the initial states of the model.");
```
Is this also true for the restart algorithm? (I am fine if this is implemented this way, but would be good to understand/clarify)

sjunges commented on src/storm/settings/modules/ModelCheckerSettings.cpp:
```diff
@@ -7,13 +7,15 @@
-namespace storm {
-namespace settings {
-namespace modules {
+#include "storm/exceptions/NotImplementedException.h"
```
why do we need these headers now?

tquatmann replied:
Thx, removed them.

sjunges commented on src/storm/modelchecker/helper/conditional/ConditionalHelper.cpp:
```diff
@@ -0,0 +1,772 @@
+    // TerminalStates is a superset of conditionStates and dom(nonZeroTargetStateValues).
+    // For a terminalState that is not a conditionState, it is impossible to (reach the condition and not reach the target).
+
+    ValueType targetValue(uint64_t state) const {
```
I know this is more c++ style than what we usually do, but would expect in this code base that this is called getTargetValue?

sjunges commented on src/storm/modelchecker/helper/conditional/ConditionalHelper.cpp:
```diff
@@ -0,0 +1,772 @@
+    auto const targetAndNotCondFailStates = extendedTargetStates & ~(extendedConditionStates | universalObservationFailureStates);
+    computeReachabilityProbabilities(env, nonZeroTargetStateValues, dir, transitionMatrix, targetAndNotCondFailStates, allStates, extendedConditionStates);
+
+    auto terminalStates = extendedConditionStates | extendedTargetStates | universalObservationFailureStates;
```
Ah, I see. Maybe swap order with terminStatesThatReachCondition

sjunges commented on src/storm/modelchecker/helper/conditional/ConditionalHelper.cpp:
```diff
@@ -0,0 +1,772 @@
+    // @see doi.org/10.1007/978-3-642-54862-8_43
+    STORM_LOG_THROW(goal.hasRelevantValues(), storm::exceptions::NotSupportedException,
+                    "No initial state given. Conditional probabilities can only be computed for models with a single initial state.");
+    STORM_LOG_THROW(goal.relevantValues().getNumberOfSetBits() == 1, storm::exceptions::NotSupportedException,
```
Side comment. I guess we check often whether the number of setBits is exaclty one, maybe we should have a method for that?

tquatmann committed:
Implemented suggestion by sjunges

- remove unnecessary include
- rename ConditionalAlgorithm to ConditionalAlgorithmSetting
- rename targetValue to getTargetValue
- reordered some assignments when computing normal form.
- elaborate on normal form precision factor

tquatmann commented on src/storm/modelchecker/helper/conditional/ConditionalAlgorithm.cpp:
```diff
@@ -0,0 +1,36 @@
+namespace storm {
+std::ostream& operator<<(std::ostream& stream, ConditionalAlgorithm const& algorithm) {
+    switch (algorithm) {
+        case ConditionalAlgorithm::Default:
```
I changed the enum to `ConditionalAlgorithmSetting`

tquatmann committed:
added `bool BitVector::hasUniqueSetBit() const`

tquatmann commented:
Thanks for the great comments! :)


I should mention that this PR right now breaks support for conditional probabilities for interval MDPs (which in general were also not handled correctly before either).

sjunges commented on src/storm/modelchecker/helper/conditional/ConditionalHelper.cpp:
```diff
@@ -0,0 +1,772 @@
+    auto analysisEnv = env;
+    if (env.solver().isForceSoundness()) {
+        normalFormConstructionEnv.solver().minMax().setPrecision(env.solver().minMax().getPrecision() *
+                                                                 storm::utility::convertNumber<storm::RationalNumber, std::string>("1/10"));
```
If there were options, I would set this hardcoded in the options, such that people can overrride the options if they want to. Not essential though.

tquatmann replied:
Yes, maybe there will be some more options introduced in the future and then I'm fine to add that.

sjunges commented:
TLDR regarding interval models: Interval MCs are fine (which was the use case). We can enable that in another PR.

LGTM
